sphinx/ext/intersphinx.py

```diff
@@ -305,9 +305,19 @@ def missing_reference(app: Sphinx, env: BuildEnvironment, node: pending_xref,
                     to_try.append((inventories.named_inventory[setname], full_qualified_name))
     for inventory, target in to_try:
         for objtype in objtypes:
-            if objtype not in inventory or target not in inventory[objtype]:
-                continue
-            proj, version, uri, dispname = inventory[objtype][target]
+            # Special case handling for term to search in a case insensitive fashion
+            if objtype == 'std:term' and objtype in inventory:
+                cased_keys = {k.lower(): k for k in inventory[objtype].keys()}
```

Comment:
We could avoid creating this whole dictionary by checking each key in turn?

```
lower_target = target.lower()
if any(lower_target == inv_key.lower() for inv_key in inventory[objtype]):
    ...
```

We could avoid doing even this if a case sensitive search returns a result?

```python
if objtype == 'std:term' and objtype in inventory:
    if target in inventory[objtype]:
        proj, version, uri, dispname = inventory[objtype][target]
    else:
        lower_target = target.lower()
        if any(lower_target == inv_key.lower() for inv_key in inventory[objtype]):
            ...
       else:
            continue
else:
    ...
```

Reply:
I think you still have to build the dict, because after determining that there *is* a case insensitive match you have to look up the target, which is inherently a case sensitive process as it's just a dict key? That or something like

```python
            if objtype not in inventory:
                continue
            if target in inventory[objtype]:
                # Case sensitive match
                proj, version, uri, dispname = inventory[objtype][target]
            elif objtype == 'std:term':
                # Check for potential case insensitive match
                target_lower = target.lower()
                insensitive_matches = list(filter(lambda k: k.lower() == target_lower, inventory[objtype].keys()))
                if insensitive_matches:
                    proj, version, uri, dispname = inventory[objtype][insensitive_matches[0]]
                else:
                    continue
            else:
                continue
```

but I'm not sure this is any more efficient. I guess it's building a list rather than a dict.

Reply:
Oh you're totally right :facepalm: This list approach will save a small amount of memory at least, plus it gives us the opportunity to raise a warning if multiple keys match.
I like the structure of the `if` statements that you're proposing too!

Reply:
I've pushed a commit with the above (plus comments). I guess the argument for doing this with a dict is if there are a lot of cases and you can move the dict generation out of the per-lookup loop, so have it outside of the `for inventory, target in to_try:` block, but you still have the memory occupancy to deal with. Priorities, as always!